internal/clients/clientimpl/baseimagematcher/baseimagematcher.go

```diff
@@ -0,0 +1,160 @@
+package baseimagematcher
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"slices"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+	"github.com/opencontainers/go-digest"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	maxConcurrentRequests = 1000
+)
+
+// OSVMatcher implements the VulnerabilityMatcher interface with a osv.dev client.
+// It sends out requests for every package version and does not perform caching.
+type DepsDevBaseImageMatcher struct {
+	Client   http.Client
+	Reporter reporter.Reporter
+}
+
+func (matcher *DepsDevBaseImageMatcher) MatchBaseImages(ctx context.Context, layerMetadata []models.LayerMetadata) ([][]models.BaseImageDetails, error) {
+	baseImagesMap := make([][]models.BaseImageDetails, len(layerMetadata))
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentRequests)
+
+	var runningDigest digest.Digest
+	for i, l := range layerMetadata {
+		if l.DiffID == "" {
+			continue
+		}
+
+		if runningDigest == "" {
+			runningDigest = l.DiffID
+		} else {
+			runningDigest = digest.FromBytes([]byte(runningDigest + " " + l.DiffID))
+		}
+
+		chainID := runningDigest
+		g.Go(func() error {
+			if ctx.Err() != nil {
+				return ctx.Err() // this value doesn't matter to errgroup.Wait(), it will be ctx.Err()
+			}
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.deps.dev/v3alpha/querycontainerimages/"+chainID.String(), nil)
+			if err != nil {
+				matcher.Reporter.Errorf("failed to build request: %s\n", err)
+				return nil
+			}
+
+			resp, err := matcher.Client.Do(req)
+			if err != nil {
+				matcher.Reporter.Errorf("deps.dev API error: %s\n", err)
+				return nil
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode == http.StatusNotFound {
+				return nil
+			}
+
+			if resp.StatusCode != http.StatusOK {
```

Comment:
nit: this function is quite long. Could we separate all the requests into a separate function to improve readability?

Reply:
Done